app/bundles/EmailBundle/Entity/StatRepository.php

```diff
@@ -744,4 +747,79 @@ public function getStatsSummaryForContacts(array $contacts): array
 
         return $contacts;
     }
+
+    /**
+     * @param array<int> $entityIds
+     *
+     * @return array<int, array<string, int|string>>
+     *
+     * @throws Exception
+     */
+    public function getStatsSummaryByCountry(\DateTimeImmutable $dateFrom, \DateTimeImmutable $dateTo, array $entityIds, string $sourceType = 'email'): array
+    {
+        $queryBuilder               = $this->getEntityManager()->getConnection()->createQueryBuilder();
+        $subQueryBuilder            = $this->getEntityManager()->getConnection()->createQueryBuilder();
+
+        $leadAlias     = 'l'; // leads
+        $statsAlias    = 'es'; // email_stats
+        $subQueryAlias = 'sq'; // sub query
+        $cutAlias      = 'cut'; // channel_url_trackables
+        $pageHitsAlias = 'ph'; // page_hits
+
+        // use sub query to get page hits for and unique page hits selected contacts
+        $subQueryBuilder->select(
+            "COUNT({$pageHitsAlias}.id) AS hits",
+            "COUNT(DISTINCT({$pageHitsAlias}.redirect_id)) AS unique_hits",
+            "{$cutAlias}.channel_id",
+            "{$pageHitsAlias}.lead_id"
+        )
+            ->from(MAUTIC_TABLE_PREFIX.'channel_url_trackables', $cutAlias)
+            ->join(
+                $cutAlias,
+                MAUTIC_TABLE_PREFIX.'page_hits',
+                $pageHitsAlias,
+                "{$cutAlias}.redirect_id = {$pageHitsAlias}.redirect_id AND {$cutAlias}.channel_id = {$pageHitsAlias}.source_id"
+            )
+            ->where("{$cutAlias}.channel = 'email' AND {$pageHitsAlias}.source = 'email'")
+            ->groupBy("{$cutAlias}.channel_id, {$pageHitsAlias}.lead_id");
+
+        // main query
+        $queryBuilder->addSelect(
+            "COUNT({$statsAlias}.id) AS `sent_count`",
+            "SUM(IF({$statsAlias}.is_read IS NULL, 0, {$statsAlias}.is_read)) AS `read_count`",
+            "SUM(IF({$subQueryAlias}.hits is NULL, 0, 1)) AS `clicked_through_count`",
+        )->from(MAUTIC_TABLE_PREFIX.'email_stats', $statsAlias)
+            ->rightJoin(
+                $statsAlias,
+                MAUTIC_TABLE_PREFIX.'leads',
+                $leadAlias,
+                "{$statsAlias}.lead_id=l.id"
+            )->leftJoin(
+                $statsAlias,
+                "({$subQueryBuilder->getSQL()})",
+                $subQueryAlias,
+                "{$statsAlias}.email_id = {$subQueryAlias}.channel_id AND {$statsAlias}.lead_id = {$subQueryAlias}.lead_id"
+            );
```

Comment:
I see no reason to split this query into two separate ones. In this way, all data is retrieved using one query, which is also not particularly complicated for the database. Why should queries be separated?

Reply:
Let's leave this once someone hits the bottleneck then.

Reply:
Splitting into two separate queries would probably require a join to the lead table in each, and this may slow down execution. I noticed that we can further narrow down the number of rows to be searched by filtering the results for campaigns not only by event ids, but also by email ids, so I made a change in generating the query for the campaign.

I analyzed the query using EXPLAIN and profiling on a database that has ~20 million rows in the email stats table. For a given period of 4 years, the query execution time was approximately 0.7s for a campaign with 90,000 e-mails sent (1 campaign event to 36 countries) and for a campaign with 50 e-mail sending events to over 80 countries it was approximately 0.13s. 

I understand that execution time depends on server conditions, database settings, amount of data and other factors, but it shouldn't be very bad. 

Query:
```
SELECT COUNT(es.id) AS `sent_count`, SUM(IF(es.is_read IS NULL, 0, es.is_read)) AS `read_count`, SUM(IF(sq.hits is NULL, 0, 1)) AS `clicked_through_count`, l.country AS `country` FROM email_stats es RIGHT JOIN leads l ON es.lead_id=l.id LEFT JOIN (SELECT COUNT(ph.id) AS hits, COUNT(DISTINCT(ph.redirect_id)) AS unique_hits, cut.channel_id, ph.lead_id FROM channel_url_trackables cut INNER JOIN page_hits ph ON cut.redirect_id = ph.redirect_id AND cut.channel_id = ph.source_id WHERE (cut.channel = 'email' AND ph.source = 'email') AND (cut.channel_id in (:emails)) GROUP BY cut.channel_id, ph.lead_id) sq ON es.email_id = sq.channel_id AND es.lead_id = sq.lead_id WHERE (es.source_id in (:events)) AND (es.source = :source) AND (es.date_sent BETWEEN :dateFrom AND :dateTo) GROUP BY l.country
```